ext/spl/spl_array.c

```diff
@@ -1544,6 +1579,16 @@ PHP_METHOD(RecursiveArrayIterator, hasChildren)
 }
 /* }}} */
 
+static void spl_instantiate_child_arg(zend_class_entry *pce, zval *retval, zval *arg1, zval *arg2) /* {{{ */
+{
+	object_init_ex(retval, pce);
+	spl_array_object *new_intern = Z_SPLARRAY_P(retval);
+	new_intern->is_child = true;
+	new_intern->bucket = (Bucket *)((char*)(arg1) - XtOffsetOf(Bucket, val));
```

Comment:
This line needs a comment, because I have no idea what the hell is happening here.

You are doing some address arithmetic but I don't really get why you are using a `Bucket` and not just a `zval` for the ``spl_array_object`` struct, as it seems you're just using the zval part anyway?

Reply:
These things happened when creating this object:
1. Setting `new_intern->is_child = true` indicating that this child object was created by `RecursiveArrayIterator::getChildren()` method.
2.  After copying the array by `zend_array_dup`, I need to share the new `array` between the parent and child object.
3. So, I need this `Bucket` that belong to parent object.  In this way, I can directly modify this value `intern->bucket->val = intern->array;`, so that changes between child object and parent object can affect each other.

Reply:
```c
new_intern->bucket = (Bucket *)((char*)(arg1) - XtOffsetOf(Bucket, val));
```
So I can get the `bucket` of the parent object by address arithmetic and replace `bucket->val` with copy `array`.